### PR TITLE
swarm: tracing to help fix `unexpected EOF`

### PR DESCRIPTION
**Pull request by nonsense**
This PR is an attempt to address: https://github.com/ethersphere/go-ethereum/issues/373

It is also adding a slight refactor: accept interfaces, but return concrete types.

### CODE REVIEW

File: swarm/storage/memstore.go [hunk 1]
```diff
@@ -55,7 +54,7 @@ func NewMemStore(params *StoreParams, _ *LDBStore) (m *MemStore) {
 	}
 
 	requestEvicted := func(key interface{}, value interface{}) {
```
Comment (nonsense) [inline swarm/storage/memstore.go]: @gbalint I am afraid this was being called a lot when I ran tests with bigger files. Even though no errors were triggered, we were working under the assumption that requests are not evicted from the cache. We should re-think this.

File: swarm/api/api.go [hunk 2]
```diff
@@ -308,7 +308,7 @@ func (self *Api) Put(content, contentType string, toEncrypt bool) (k storage.Key
 // Get uses iterative manifest retrieval and prefix matching
 // to resolve basePath to content using dpa retrieve
 // it returns a section reader, mimeType, status and an error
-func (self *Api) Get(key storage.Key, path string) (reader storage.LazySectionReader, mimeType string, status int, err error) {
+func (self *Api) Get(key storage.Key, path string) (reader *storage.LazyChunkReader, mimeType string, status int, err error) {
```
Comment (gbalint) [inline swarm/api/api.go:311]: Maybe I miss something, but I don't get why you changed from `LazySectionReader` interface to `LazyChunkReader` struct here and in the other places you modified.

Comment (nonsense) [reply swarm/api/api.go:311]: http://idiomaticgo.com/post/best-practice/accept-interfaces-return-structs/ - TLDR - easier to read the code.

**Review by gbalint — APPROVED** [review]

File: swarm/api/api.go (hunk 2 above)
Comment (nonsense) [reply swarm/api/api.go:311]: @janos yes, we are no longer using the Joiner interface and it has been removed. If we need it, then we fix this code to return an interface.

**Review by janos — APPROVED** [review]
LGTM

**Pull request merged**